README.md
# gnfinder

Ruby gem to access functionality of [GNfinder] project written in Go. This gem
allows to perform fast and accurate scientific name finding in texts,
web-pages, as well as a large variety of documents. Document files can be
accessed either locally or via a URL.


<!-- vim-markdown-toc GFM -->

* [Requirements](#requirements)
* [Installation](#installation)
* [Usage](#usage)
  * [Finding names in a text using default settings](#finding-names-in-a-text-using-default-settings)
  * [Finding names by a URL](#finding-names-by-a-url)
  * [Finding names in a file](#finding-names-in-a-file)
  * [Optionally disable Bayes search](#optionally-disable-bayes-search)
  * [Set a language for the text](#set-a-language-for-the-text)
* [Set automatic detection of text's language](#set-automatic-detection-of-texts-language)
  * [Set verification option](#set-verification-option)
  * [Set preferred data-sources list](#set-preferred-data-sources-list)
  * [Combination of parameters.](#combination-of-parameters)
* [Development](#development)

<!-- vim-markdown-toc -->

## Requirements

This gem uses REST API to access a running [GNfinder] server. You can find how
to run it in [GNfinder] README file. By default it uses
`https://gnfinder.globalnames.org/api/v0`

## Installation

```bash
gem install gnfinder
```

## Usage

The purpose of this gem is to access [GNfinder] functionality from Ruby
applications. If you need to find names using other languages, use the
[source code][client] of this gem for reference. For other usages read
the original Go-lang [GNfinder] README file.

First you need to create an instance of a `gnfinder` client

```ruby
require 'gnfinder'

gf = Gnfinder::Client.new
```

By default the client will try to connect to
`https://gnfinder.globalnames.org/api/v0`. If you have another location for the
server use:

```ruby
require 'gnfinder'

# you can use global public gnfinder server
# located at finder-rpc.globalnames.org
gf = Gnfinder::Client.new(host = 'finder.example.org', port = 80)

# localhost, port 8000
gf = Gnfinder::Client.new(host = '0.0.0.0', port = 8000)
```

### Finding names in a text using default settings

You can find format of returning result in [GNfinder API docs]

```ruby
txt = File.read('utf8-text-with-names.txt')

res = gf.find_names(txt)
puts res.names[0].value
puts res.names[0].odds
```

### Finding names by a URL

If you need to find names in an HTML page, or a PDF document available on
Internet, use `find_url` method.

```ruby
url = 'https://en.wikipedia.org/wiki/Monochamus_galloprovincialis'
res = gf.find_url(url)
puts res.names[0].value
puts res.names[0].odds
```

### Finding names in a file

Many different file types are supported (PDF, JPB, TIFF, MS Word, MS Excel
etc).

```ruby
path = "/path/to/file.pdf"
res = gf.find_file(path)
puts res.names[0].value
```

Support of file-uploading uses 'multipart/form' approach. Here is an
illustration for `curl`:

```bash
curl -v -F sources[]=1 -F sources[]=12 -F file=@file.pdf \
    https://finder.globalnames.org/api/v0/find
```

Returned result is quite detailed and contains many accessor methods, for
example:

  * value: name-string cleaned up for verification.
  * verbatim: name-string as it was found in the text.
  * odds: Bayes' odds value. For example odds 0.1 would mean that according to
    the algorithm there is 1 chance out of 10 that the name-string is
    a scientific name. This field will be empty if Bayes algorithms did not run.

### Optionally disable Bayes search

Some languages that are close to Latin (Italian, French, Portugese) would
generate too many false positives. To decrease amount of false positives you
can disable Bayes algorithm by running:

```ruby
names = gf.find_names(txt, no_bayes: true).names
```

### Set a language for the text

It is possible to supply the prevalent language to set a language for a text
by hand. That might Bayes algorithms work better

List of supported languages will increase with time.

```ruby
res = gf.find_names(txt, language: 'eng')
puts res.language
res = gf.find_names(txt, language: 'deu')
puts res.language

# Setting is ignored if language string is not known by gnfinder.
# Only 3-character notations iso-639-2 code are supported
res = gf.find_names(txt, language: 'rus')
puts res.language
```
## Set automatic detection of text's language

To enable automatic detection of prevalent language of a text use:

res = gf.find_names(txt, detect_language: true)
puts res.language
puts res.detect_language
puts res.language_detected

If detected language is not yet supported by Bayes algorithm, default
language (English) will be used.

### Set verification option

In case if found names need to be validated against a large collection of
name-strings, use `with_verification` option. For each name algorithm will
return the following information:

  * match type:
    -	``NONE``: name-string is unknown
    - ``EXACT``: name-string matched exactly.
    - ``CANONICAL_EXACT``: canonical form of a name-string matched exactly.
    - ``CANONICAL_FUZZY``: fuzzy match of a canonical string.
    - ``PARTIAL_EXACT``: only part of a name matched. For examle only genus of a
      species was found.
    - ``PARTIAL_FUZZY``: fuzzy match of a partial result. For example canonical
      form of a trinomial matched on a species level with some corrections.
  * source_id: ID of a data-source of a best matched result. Data source IDs
    can be compared with the [data-source list]
  * curated: true if name-string was found in some data-sources that are
    deemed to curated by humans.
  * path: the classification path of a matched name (if available)

```ruby
res = gf.find_names(txt, verification: true)
```

### Set preferred data-sources list

Sometimes it is important to know if a name was found in a particular
data-source (data-sources). There is a parameter that takes IDs from the
[data-source list]. If a name-string was found in these data-sources, match
results will be returned back.

```ruby
res = gf.find_names(txt, verification: true, sources: [1, 4, 179])
```
### Combination of parameters.

It is possible to combine parameters. However if a parameter makes no sense in
a particular context it is silently ignored.

```ruby
# Runs Bayes' algorithms using English training set, runs verification and
# returns matched results for 3 data-sources if they are available.
res = gf.find_names(txt, language: 'eng', verification: true,
                           sources: [1, 4, 179])

# Ignores `sources:` settings, because `with_verification` is not set to `true`
res = gf.find_names(txt, language: 'eng', sources: [1, 4, 179])
```

## Development

If you get an error, you might need to set a ``GOPATH`` environment variable.

After starting the server with default host and port (localhost:8778) you will
be able to run tests for this Ruby client with:

```bash
bundle exec rake
```

To run rubocop test

```bash
bundle exec rake rubocop
```

To run tests without rubocop
```bash
bundle exec rspec
```

[GNfinder]: https://github.com/gnames/gnfinder
[gnfinder recent release]: https://github.com/gnames/gnfinder/releases
[Go]: https://golang.org/doc/install
[client]: https://github.com/GlobalNamesArchitecture/gnfinder/blob/master/lib/gnfinder/client.rb
[data-source list]: http://index.globalnames.org/datasource
[tests]: https://github.com/GlobalNamesArchitecture/gnfinder/blob/master/spec/lib/client_spec.rb
[GNfinder API docs]: https://apidoc.globalnames.org/gnfinder


CHANGELOG.md
# Changelog

## Unreleased

## [v0.18.0]

- Bring tests in accordance to API changes in v0.18.0

## [v0.17.0]

- Update to gnfinder v0.17.0

## [v0.15.5]

- Add [#16]: Search names using URL.

## [v0.15.4]

- Add [#14]: Compatibility with gnfider v0.15.4, switching to REST API.

## [v0.2.0]

- Compatibility with gnfinder v0.7.0

## [v0.1.3]

- Fix [#8]: gRPC output is similar to CLI output for gnfinder.

## [v0.1.2]

- Fix [#7]: utf-8 diacritics support in input text fails.

## [v0.1.1]

- Add [#6]: return start and end of names-strings.

## [v0.1.0]

- Add [#5]: add an optional list of preferred data-sources for verification.
- Add [#4]: add an option for verification of found names.
- Add [#3]: enable sending options to gRPC service.
- Add [#2]: send a text and receive found names.
- Add [#1]: connect to gRPC server of `gnfinder` Go-lang program.

## Footnotes

This document follows [changelog guidelines]

[v0.18.0]: https://github.com/GlobalNamesArchitecture/gnfinder/compare/v0.17.0...v0.18.0
[v0.17.0]: https://github.com/GlobalNamesArchitecture/gnfinder/compare/v0.15.5...v0.17.0
[v0.15.5]: https://github.com/GlobalNamesArchitecture/gnfinder/compare/v0.15.4...v0.15.5
[v0.15.4]: https://github.com/GlobalNamesArchitecture/gnfinder/compare/v0.2.0...v0.15.4
[v0.2.0]: https://github.com/GlobalNamesArchitecture/gnfinder/compare/v0.1.3...v0.2.0
[v0.1.3]: https://github.com/GlobalNamesArchitecture/gnfinder/compare/v0.1.2...v0.1.3
[v0.1.2]: https://github.com/GlobalNamesArchitecture/gnfinder/compare/v0.1.1...v0.1.2
[v0.1.1]: https://github.com/GlobalNamesArchitecture/gnfinder/compare/v0.1.0...v0.1.1
[v0.1.0]: https://github.com/GlobalNamesArchitecture/gnfinder/tree/v0.1.0

[#20]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/20
[#19]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/19
[#18]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/18
[#17]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/17
[#16]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/16
[#15]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/15
[#14]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/14
[#13]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/13
[#12]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/12
[#11]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/11
[#10]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/10
[#9]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/9
[#8]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/8
[#7]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/7
[#6]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/6
[#5]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/5
[#4]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/4
[#3]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/3
[#2]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/2
[#1]: https://github.com/GlobalNamesArchitecture/gnfinder/issues/1


[changelog guidelines]: https://github.com/olivierlacan/keep-a-changelog


spec/spec_helper.rb
# frozen_string_literal: true

require 'rspec'
require 'gnfinder'


spec/lib/client_spec.rb
# frozen_string_literal: true

# Gnfinder is a namespace module for gndinfer gem.
describe Gnfinder::Client do
  let(:subject) { Gnfinder::Client.new }
  # let(:subject) { Gnfinder::Client.new(host = '0.0.0.0', port = 8080) }

  describe '#ping' do
    it 'connects to the server' do
      expect(subject.ping).to eq 'pong'
    end
  end

  describe '#version' do
    it 'returns version of Go gnfinder' do
      expect(subject.gnfinder_version.version).to match(/^v\d+\.\d+\.\d+/)
    end
  end

  describe '#find_file' do
    dir = File.absolute_path(__dir__)
    path = File.join(dir, '..', 'files', 'test.txt')

    it 'returns list of name_strings' do
      names = subject.find_file(path).names
      expect(names[0].name).to eq 'Monochamus galloprovincialis'
      expect(names[0].cardinality).to eq 2
    end

    it 'works with images' do
      img = File.join(dir, '..', 'files', 'image.jpg')
      names = subject.find_file(img).names
      expect(names[0].name).to eq 'Baccha'
      expect(names[0].cardinality).to eq 1
    end

    it 'works with images' do
      pdf = File.join(dir, '..', 'files', 'file.pdf')
      names = subject.find_file(pdf).names
      expect(names[0].name).to eq 'Passiflora'
      expect(names[0].cardinality).to eq 1
    end

    it 'supports no_bayes option' do
      names1 = subject.find_file(path).names
      opts = { no_bayes: true }
      names2 = subject.find_file(path, opts).names
      expect(names1.size).to be > names2.size
    end

    it 'supports language option' do
      res = subject.find_file(path)
      expect(res.language).to eq 'eng'
      expect(res.language_detected).to eq nil
      expect(res.with_language_detection).to be false

      opts = { language: 'deu' }
      res = subject.find_file(path, opts)
      expect(res.language).to eq 'deu'
      expect(res.language_detected).to eq nil
      expect(res.with_language_detection).to be false
    end

    it 'silently ignores unknown language' do
      res = subject.find_file(path)
      expect(res.language).to eq 'eng'
      expect(res.language_detected).to eq nil
      expect(res.with_language_detection).to be false

      opts = { language: 'whatisit' }
      res = subject.find_file(path, opts)
      expect(res.language).to eq 'eng'
      expect(res.language_detected).to eq nil
      expect(res.with_language_detection).to be false
    end

    it 'can detect language' do
      opts = { language: 'detect' }
      res = subject.find_file(path, opts)
      expect(res.language).to eq 'eng'
      expect(res.language_detected).to eq 'eng'
      expect(res.with_language_detection).to be true
    end

    it 'supports tokens around option' do
      opts = { words_around: 2 }
      names = subject.find_file(path, opts).names
      expect(names[2].words_before).to eq %w[of beetle]
      expect(names[2].words_after).to eq %w[image Monochamus]
    end

    it 'supports verification option' do
      opts = { verification: true }
      names = subject.find_file(path, opts).names
      expect(names[0].verification.best_result.match_type).to eq 'Exact'
      expect(names[0].verification.best_result.matched_cardinality).to eq 2
    end

    it 'supports verification with sources' do
      opts = { sources: [1] }
      names = subject.find_file(path, opts).names
      expect(names[0].verification.results)
        .to be_nil
      expect(names[0].verification.best_result.data_source_title_short)
        .to eq 'Catalogue of Life'
    end

    it 'returns the position of a name in a text' do
      names = subject.find_file(path).names
      expect(names[0].start).to eq 15
      expect(names[0].end).to eq 43
    end

    it 'breaks on wrong paths' do
      expect { subject.find_file('wrong/path') }.to raise_error(/No such file/)
    end
  end

  describe '#find_url' do
    it 'gets metadata about results' do
      res = subject.find_url('https://example.com')
      expect(res.words_around).to eq 0
      opts = { words_around: 2 }
      res = subject.find_url('https://example.com', opts)
      expect(res.words_around).to eq 2
    end

    it 'returns list of name_strings' do
      names = subject.find_url('https://en.wikipedia.org/wiki/Monochamus_galloprovincialis').names
      expect(names[0].name).to eq 'Monochamus galloprovincialis'
      expect(names[0].cardinality).to eq 2
    end
  end

  describe '#find_names' do
    it 'gets metadata about results' do
      res = subject.find_names('Pardosa moesta is a spider')
      expect(res.words_around).to eq 0
      opts = { words_around: 2 }
      res = subject.find_names(
        'It is very interesting that Pardosa moesta is a spider', opts
      )
      expect(res.words_around).to eq 2
    end

    it 'returns list of name_strings' do
      names = subject.find_names('Pardosa moesta is a spider').names
      expect(names[0].name).to eq 'Pardosa moesta'
      expect(names[0].verbatim).to eq 'Pardosa moesta'
      expect(names[0].cardinality).to eq 2
    end

    it 'finds nomenclatural annotation for a name' do
      names = subject.find_names('Pardosa moesta sp. n. is a spider').names
      expect(names[0].name).to eq 'Pardosa moesta'
      expect(names[0].annotation_nomen).to eq 'sp. n.'
      expect(names[0].annotation_nomen_type).to eq 'SP_NOV'
    end

    it 'supports no_bayes option' do
      names = subject.find_names('Pardosa moesta is a spider').names
      expect(names[0].odds_log10).to be > 2.0
      names = subject.find_names(
        'Falsificus erundiculus var. pridumalus is a spider'
      ).names
      expect(names.size).to eq 1

      opts = { no_bayes: true }
      names = subject.find_names('Pardosa moesta is a spider', opts).names
      expect(names[0].odds_log10).to eq nil
      names = subject.find_names(
        'Falsificus erundiculus var. pridumalus is a spider', opts
      ).names
      expect(names.size).to eq 0
    end

    it 'supports language option' do
      res = subject.find_names('Pardosa moesta is a spider')
      expect(res.language).to eq 'eng'
      expect(res.language_detected).to eq nil
      expect(res.with_language_detection).to be false

      opts = { language: 'deu' }
      res = subject.find_names('Pardosa moesta is a spider', opts)
      expect(res.language).to eq 'deu'
      expect(res.language_detected).to eq nil
      expect(res.with_language_detection).to be false
    end

    it 'silently ignores unknown language' do
      res = subject.find_names('Pardosa moesta is a spider')
      expect(res.language).to eq 'eng'
      expect(res.language_detected).to eq nil
      expect(res.with_language_detection).to be false

      opts = { language: 'whatisit' }
      res = subject.find_names('Pardosa moesta is a spider', opts)
      expect(res.language).to eq 'eng'
      expect(res.language_detected).to eq nil
      expect(res.with_language_detection).to be false
    end

    it 'supports detect_language option' do
      res = subject.find_names('Pardosa moesta is a spider')
      expect(res.language).to eq 'eng'
      expect(res.language_detected).to eq nil
      expect(res.with_language_detection).to be false

      opts = { language: 'detect' }
      res = subject.find_names(
        'Pardosa moesta это латинское название одного паука', opts
      )
      expect(res.language).to eq 'eng'
      expect(res.language_detected).to eq 'rus'
      expect(res.with_language_detection).to be true

      opts = { language: 'detect' }
      res = subject.find_names(
        'Pardosa moesta ist ein lateinischer Name für eine kleine Spinne', opts
      )
      expect(res.language).to eq 'deu'
      expect(res.language_detected).to eq 'deu'
      expect(res.with_language_detection).to be true

      opts = { language: 'detect' }
      res = subject.find_names(
        'Pardosa moesta это латинское название одного паука', opts
      )
      expect(res.language).to eq 'eng'
      expect(res.language_detected).to eq 'rus'
      expect(res.with_language_detection).to be true
    end

    it 'supports tokens around option' do
      opts = { words_around: 2 }
      names = subject.find_names(
        'It is very interesting that Pardosa moesta is a spider', opts
      ).names
      expect(names[0].words_before).to eq %w[interesting that]
      expect(names[0].words_after).to eq %w[is a]
    end

    it 'supports verification option' do
      opts = { verification: true }
      names = subject.find_names('Pardosa moesta is a spider', opts).names
      expect(names[0].verification.best_result.match_type).to eq 'Exact'
      expect(names[0].verification.best_result.matched_cardinality).to eq 2
    end

    it 'supports verification with sources' do
      opts = { verification: true, sources: [1] }
      res = subject.find_names('Pardosa moesta is a spider', opts)
      names = res.names
      expect(names[0].verification.results)
        .to be_nil
      expect(names[0].verification.best_result.data_source_title_short)
        .to eq 'Catalogue of Life'
    end

    it 'returns the position of a name in a text' do
      names = subject.find_names('Pardosa moesta is a spider').names
      expect(names[0].start).to eq 0
      expect(names[0].end).to eq 14
    end

    it 'works with utf8 text' do
      names = subject.find_names('Pedicia apusenica (Ujvárosi and Starý 2003)')
                     .names
      expect(names[0].name).to eq 'Pedicia apusenica'
    end

    it 'works with empty text' do
      names = subject.find_names('').names
      expect(names.size).to eq 0
    end

    it 'gets metadata' do
      res = subject.find_names('Pardosa moesta is a very interesting spider')
      expect(res.date).to match(/\d{4}/)
      expect(res.language).to eq 'eng'
      expect(res.language_detected).to eq nil
      expect(res.with_language_detection).to eq false
      expect(res.total_words).to be 7
      expect(res.total_name_candidates).to be 1
      expect(res.total_names).to be 1
    end

    it 'gets metadata with language option' do
      opts = { language: 'deu' }
      res = subject
            .find_names('Pardosa moesta is a very interesting spider', opts)
      expect(res.language_detected).to eq nil
      expect(res.with_language_detection).to eq false
      expect(res.language).to eq 'deu'
    end

    it 'ignores language option if language string is unknown' do
      opts = { language: 'German' }
      res = subject
            .find_names('Pardosa moesta is a very interesting spider', opts)
      expect(res.gnfinder_version).to match(/^v\d+\.\d+\.\d+/)
      expect(res.language_detected).to eq nil
      expect(res.with_language_detection).to eq false
      expect(res.language).to eq 'eng'
    end
  end
end


spec/lib/gnfinder_spec.rb
# frozen_string_literal: true

describe Gnfinder do
  describe '.version' do
    it 'shows version' do
      expect(subject.version).to match(/\d+\.\d+.\d+/)
    end
  end
end


lib/hash.rb
# frozen_string_literal: true

# Monkey patch for underscores.
class String
  def underscore
    gsub(/::/, '/')
      .gsub(/([a-z\d])([A-Z])/) do
      "#{Regexp.last_match(1)}_#{Regexp.last_match(2)}"
    end
      .downcase
  end
end

# Hash monkey patches from ActiveSupport
class Hash
  # Returns a new hash with all keys converted to strings.
  #
  #   hash = { name: 'Rob', age: '28' }
  #
  #   hash.stringify_keys
  #   # => {"name"=>"Rob", "age"=>"28"}
  def stringify_keys
    transform_keys(&:to_s)
  end

  # Destructively converts all keys to strings. Same as
  # +stringify_keys+, but modifies +self+.
  def stringify_keys!
    transform_keys!(&:to_s)
  end

  # Returns a new hash with all keys converted to symbols, as long as
  # they respond to +to_sym+.
  #
  #   hash = { 'name' => 'Rob', 'age' => '28' }
  #
  #   hash.symbolize_keys
  #   # => {:name=>"Rob", :age=>"28"}
  def symbolize_keys
    transform_keys do |key|
      key.to_sym
    rescue StandardError
      key
    end
  end
  alias to_options symbolize_keys

  # Destructively converts all keys to symbols, as long as they respond
  # to +to_sym+. Same as +symbolize_keys+, but modifies +self+.
  def symbolize_keys!
    transform_keys! do |key|
      key.to_sym
    rescue StandardError
      key
    end
  end
  alias to_options! symbolize_keys!

  # Validates all keys in a hash match <tt>*valid_keys</tt>, raising
  # +ArgumentError+ on a mismatch.
  #
  # Note that keys are treated differently than HashWithIndifferentAccess,
  # meaning that string and symbol keys will not match.
  def assert_valid_keys(*valid_keys)
    valid_keys.flatten!
    each_key do |k|
      unless valid_keys.include?(k)
        raise ArgumentError, "Unknown key: #{k.inspect}. Valid keys are: #{valid_keys.map(&:inspect).join(', ')}"
      end
    end
  end

  # Returns a new hash with all keys converted by the block operation.
  # This includes the keys from the root hash and from all
  # nested hashes and arrays.
  #
  #  hash = { person: { name: 'Rob', age: '28' } }
  #
  #  hash.deep_transform_keys{ |key| key.to_s.upcase }
  #  # => {"PERSON"=>{"NAME"=>"Rob", "AGE"=>"28"}}
  def deep_transform_keys(&block)
    _deep_transform_keys_in_object(self, &block)
  end

  # Destructively converts all keys by using the block operation.
  # This includes the keys from the root hash and from all
  # nested hashes and arrays.
  def deep_transform_keys!(&block)
    _deep_transform_keys_in_object!(self, &block)
  end

  # Returns a new hash with all keys converted to strings.
  # This includes the keys from the root hash and from all
  # nested hashes and arrays.
  #
  #   hash = { person: { name: 'Rob', age: '28' } }
  #
  #   hash.deep_stringify_keys
  #   # => {"person"=>{"name"=>"Rob", "age"=>"28"}}
  def deep_stringify_keys
    deep_transform_keys(&:to_s)
  end

  # Destructively converts all keys to strings.
  # This includes the keys from the root hash and from all
  # nested hashes and arrays.
  def deep_stringify_keys!
    deep_transform_keys!(&:to_s)
  end

  # Returns a new hash with all keys converted to symbols, as long as
  # they respond to +to_sym+. This includes the keys from the root hash
  # and from all nested hashes and arrays.
  #
  #   hash = { 'person' => { 'name' => 'Rob', 'age' => '28' } }
  #
  #   hash.deep_symbolize_keys
  #   # => {:person=>{:name=>"Rob", :age=>"28"}}
  def deep_symbolize_keys
    deep_transform_keys do |key|
      key.to_sym
    rescue StandardError
      key
    end
  end

  # Destructively converts all keys to symbols, as long as they respond
  # to +to_sym+. This includes the keys from the root hash and from all
  # nested hashes and arrays.
  def deep_symbolize_keys!
    deep_transform_keys! do |key|
      key.to_sym
    rescue StandardError
      key
    end
  end

  private

  # Support methods for deep transforming nested hashes and arrays.
  def _deep_transform_keys_in_object(object, &block)
    case object
    when Hash
      object.each_with_object(self.class.new) do |(key, value), result|
        result[yield(key)] = _deep_transform_keys_in_object(value, &block)
      end
    when Array
      object.map { |e| _deep_transform_keys_in_object(e, &block) }
    else
      object
    end
  end

  # rubocop:disable all
  def _deep_transform_keys_in_object!(object, &block)
    case object
    when Hash
      object.each_key do |key|
        value = object.delete(key)
        object[yield(key)] = _deep_transform_keys_in_object!(value, &block)
      end
      object
    when Array
      object.map! { |e| _deep_transform_keys_in_object!(e, &block) }
    else
      object
    end
  end
  # rubocop:enable all
end


lib/gnfinder/client.rb
# frozen_string_literal: true

module Gnfinder
  # Gnfinder::Client connects to gnfinder server
  class Client
    def initialize(host = 'https://gnfinder.globalnames.org', port = '')
      api_path = '/api/v0'
      url = host + api_path
      url = "#{host}:#{port}#{api_path}" if port.to_s != ''
      @site = RestClient::Resource.new(url, read_timeout: 60)
      @port = port
    end

    def gnfinder_version
      resp = @site['/version'].get
      ver = JSON.parse(resp.body, symbolize_names: true)
      OpenStruct.new(ver)
    end

    def ping
      @site['/ping'].get.body
    end

    def find_file(path, opts = {})
      params = {}
      update_params(params, opts)
      file = File.new(path, 'rb')
      params = params.merge(file: file)
      resp = @site['find'].post(params)
      prepare_result(resp)
    end

    def find_url(url, opts = {})
      return to_open_struct({ "names": [] }) if url.to_s.strip == ''

      params = { url: url }
      find(params, opts)
    end

    def find_names(text, opts = {})
      return to_open_struct({ "names": [] }) if text.to_s.strip == ''

      params = { text: text }
      find(params, opts)
    end

    private

    # rubocop:disable all
    def find(params, opts = {})
      update_params(params, opts)

      resp = @site['find'].post params.to_json, {content_type: :json, accept: :json}
      prepare_result(resp)
    end
    # rubocop:enable all

    def prepare_result(response)
      output = JSON.parse(response.body)
      res = output['metadata']
      res['names'] = output['names'] || []
      res = res.deep_transform_keys(&:underscore)
      res['names'] = [] if res['names'].nil?
      to_open_struct(res)
    end

    # rubocop:disable all
    def update_params(params, opts)
      params[:noBayes] = true if opts[:no_bayes]
      params[:oddsDetails] = true if opts[:odds_details]
      params[:language] = opts[:language] if opts[:language].to_s.strip != ''

      params[:wordsAround] = opts[:words_around] if opts[:words_around] && opts[:words_around].positive?

      params[:verification] = true if opts[:verification]

      params[:sources] = opts[:sources] if opts[:sources] && !opts[:sources].empty?
    end
    # rubocop:enable all

    def to_open_struct(obj)
      case obj
      when Hash
        OpenStruct.new(obj.transform_values { |val| to_open_struct(val) })
      when Array
        obj.map { |o| to_open_struct(o) }
      else # Assumed to be a primitive value
        obj
      end
    end
  end
end


lib/protob_pb.rb
# Generated by the protocol buffer compiler.  DO NOT EDIT!
# source: protob.proto

require 'google/protobuf'

Google::Protobuf::DescriptorPool.generated_pool.build do
  add_file("protob.proto", :syntax => :proto3) do
    add_message "protob.Void" do
    end
    add_message "protob.Pong" do
      optional :value, :string, 1
    end
    add_message "protob.Version" do
      optional :version, :string, 1
      optional :build, :string, 2
    end
    add_message "protob.Params" do
      optional :text, :string, 1
      optional :no_bayes, :bool, 2
      optional :language, :string, 3
      optional :detect_language, :bool, 4
      optional :verification, :bool, 5
      optional :tokens_around, :int32, 6
      repeated :sources, :int32, 7
    end
    add_message "protob.Output" do
      optional :date, :string, 1
      optional :finder_version, :string, 2
      repeated :approach, :enum, 3, "protob.Approach"
      optional :tokens_around, :int32, 4
      optional :language, :string, 5
      optional :language_detected, :string, 6
      optional :detect_language, :bool, 7
      optional :total_tokens, :int32, 8
      optional :total_candidates, :int32, 9
      optional :total_names, :int32, 10
      repeated :names, :message, 11, "protob.NameString"
    end
    add_message "protob.NameString" do
      optional :cardinality, :int32, 1
      optional :verbatim, :string, 3
      optional :name, :string, 4
      optional :odds, :float, 5
      optional :annot_nomen, :string, 6
      optional :annot_nomen_type, :enum, 7, "protob.AnnotNomenType"
      optional :offset_start, :int32, 8
      optional :offset_end, :int32, 9
      repeated :words_before, :string, 10
      repeated :words_after, :string, 11
      optional :verification, :message, 12, "protob.Verification"
    end
    add_message "protob.Verification" do
      optional :best_result, :message, 1, "protob.ResultData"
      repeated :preferred_results, :message, 2, "protob.ResultData"
      optional :data_sources_num, :int32, 3
      optional :data_source_quality, :string, 4
      optional :retries, :int32, 5
      optional :error, :string, 6
    end
    add_message "protob.ResultData" do
      optional :data_source_id, :int32, 1
      optional :data_source_title, :string, 2
      optional :taxon_id, :string, 3
      optional :matched_name, :string, 4
      optional :matched_cardinality, :int32, 5
      optional :matched_canonical_simple, :string, 6
      optional :matched_canonical_full, :string, 7
      optional :current_name, :string, 8
      optional :current_cardinality, :int32, 9
      optional :current_canonical_simple, :string, 10
      optional :current_canonical_full, :string, 11
      optional :synonym, :bool, 12
      optional :classification_path, :string, 13
      optional :classification_rank, :string, 14
      optional :classification_ids, :string, 15
      optional :edit_distance, :int32, 16
      optional :stem_edit_distance, :int32, 17
      optional :match_type, :enum, 18, "protob.MatchType"
    end
    add_enum "protob.MatchType" do
      value :NONE, 0
      value :EXACT, 1
      value :FUZZY, 2
      value :PARTIAL_EXACT, 3
      value :PARTIAL_FUZZY, 4
    end
    add_enum "protob.AnnotNomenType" do
      value :NO_ANNOT, 0
      value :SP_NOV, 1
      value :COMB_NOV, 2
      value :SUBSP_NOV, 3
    end
    add_enum "protob.Approach" do
      value :HEURISTIC, 0
      value :BAYES, 1
    end
  end
end

module Protob
  Void = ::Google::Protobuf::DescriptorPool.generated_pool.lookup("protob.Void").msgclass
  Pong = ::Google::Protobuf::DescriptorPool.generated_pool.lookup("protob.Pong").msgclass
  Version = ::Google::Protobuf::DescriptorPool.generated_pool.lookup("protob.Version").msgclass
  Params = ::Google::Protobuf::DescriptorPool.generated_pool.lookup("protob.Params").msgclass
  Output = ::Google::Protobuf::DescriptorPool.generated_pool.lookup("protob.Output").msgclass
  NameString = ::Google::Protobuf::DescriptorPool.generated_pool.lookup("protob.NameString").msgclass
  Verification = ::Google::Protobuf::DescriptorPool.generated_pool.lookup("protob.Verification").msgclass
  ResultData = ::Google::Protobuf::DescriptorPool.generated_pool.lookup("protob.ResultData").msgclass
  MatchType = ::Google::Protobuf::DescriptorPool.generated_pool.lookup("protob.MatchType").enummodule
  AnnotNomenType = ::Google::Protobuf::DescriptorPool.generated_pool.lookup("protob.AnnotNomenType").enummodule
  Approach = ::Google::Protobuf::DescriptorPool.generated_pool.lookup("protob.Approach").enummodule
end
